ir/function.cpp

```diff
@@ -428,5 +438,127 @@ void DomTree::printDot(std::ostream &os) const {
   os << "}\n";
 }
 
+
+void LoopAnalysis::getDepthFirstSpanningTree() {
+  std::map<const BasicBlock*, std::vector<const BasicBlock*>> edges;
+  unsigned current = 0;
+  for (auto [src, dst, instr] : cfg) {
+    edges[&src].push_back(&dst);
+  }
+
+  std::function<void(const BasicBlock *)> dfs = [&](const BasicBlock *bb) {
+    node.push_back(bb);
+    number[bb] = current++;
+    for (auto tgt : edges [bb]) {
+      if (number.find(tgt) == number.end()) {
+        dfs(tgt);
+      }
+    }
+    last[number[bb]] = current - 1;
+  };
+
+  dfs(&f.getFirstBB());
+}
+
+void LoopAnalysis::printNodeType(ostream &os, const NodeType &t) const {
+  switch (t) {
+  case nonheader:   os<<"n."; break;
+  case self:        os<<"s."; break;
+  case reducible:   os<<"r."; break;
+  case irreducible: os<<"i."; break;
+  }
+}
+
+// Implemention of Tarjan-Havlak algorithm.
+//
+// Irreducible loops are partially supported yet.
+//
+// Tarjan, R. (1974). Testing Flow Graph Reducibility.
+// Havlak, P. (1997). Nesting of reducible and irreducible loops.
+void LoopAnalysis::analysis() {
+  getDepthFirstSpanningTree();
+  unsigned bb_count = f.getBBs().size();
+
+  auto isAncestor = [this](unsigned w, unsigned v) -> bool {
+    return w <= v && v <= last.at(w);
+  };
+
+  vector<set<unsigned>> nonBackPreds, backPreds;
+
+  for (unsigned w = 0; w < bb_count; ++w) {
+    nonBackPreds.push_back({});
+    backPreds.push_back({});
+    header.push_back(0);
+    type.push_back(NodeType::nonheader);
+    for (auto [src, dst, instr] : cfg) {
+      (void) instr;
+      if (&dst == node[w]) {
+        unsigned v = number.at(&src);
+        if (isAncestor(w, v))
+          backPreds[w].insert(v);
+        else
+          nonBackPreds[w].insert(v);
+      }
+    }
+  }
+  UnionFind uf(bb_count);
+
+  for (unsigned w = bb_count - 1; w != static_cast<unsigned>(-1); --w) {
+    set<unsigned> P;
+    for (unsigned v : backPreds[w])
+      if (v != w)
+        P.insert(uf.find(v));
+      else
+        type[w] = NodeType::self;
+
+    if (!P.empty()) type[w] = NodeType::reducible;
+    set<unsigned> workList(P);
+
+    while (!workList.empty()) {
+      unsigned x = *workList.begin();
+      workList.erase(x);
+
+      for (unsigned y : nonBackPreds[x]) {
+        unsigned yy = uf.find(y);
+        if (!isAncestor(w, yy)) {
+          type[w] = NodeType::irreducible;
+          nonBackPreds[w].insert(yy);
+        } else if (yy != w && P.find(yy) == P.end()) {
+          P.insert(yy);
+          workList.insert(yy);
+        }
+      }
+    }
+
+    for (unsigned x : P) {
+      header[x] = w;
+      uf.unite(x, w);
+    }
+  }
+}
+
+void LoopAnalysis::printDot(ostream &os) const {
+  os << "digraph {\n";
+  for (unsigned i = 0; i < f.getBBs().size(); ++i) {
+    if (type[i] == NodeType::nonheader)
+      continue;
+    os << '"' << bb_dot_name(node[i]->getName())
+       << "\" [xlabel=\"";
+    printNodeType(os, type[i]);
```

Comment:
I think this node type isn't need anymore. it looks ugly anyway and they are all of equal type.